src/convert/types.ts

```diff
@@ -101,7 +101,40 @@ export type ConvertResult = {
    */
   zipBuffer?: Buffer;
   /**
-   * Converted source components. Not set if archving the package.
+   * Converted source components. Not set if archiving the package.
    */
   converted?: SourceComponent[];
 };
+
+/** Stored by file on SourceComponent for stream processing */
+export type MarkedReplacement = {
+  toReplace: string | RegExp;
+  replaceWith: string;
+  singleFile?: boolean;
+};
+
+// TODO: what's the right way to get this into core/sfdxProjectJson
+export type ReplacementConfig = ReplacementLocation &
+  ReplacementSource &
+  ReplacementTarget & {
+    /** Only do the replacement if ALL of the environment values in this array match */
+    replaceWhenEnv?: [
+      {
+        env: string;
+        value: string | number | boolean;
+      }
+    ];
+  };
+
+type ReplacementLocation = { filename: string; glob?: never } | { filename?: never; glob: string };
+type ReplacementSource =
+  | { replaceWithEnv: string; replaceWithFile?: never }
+  | { replaceWithEnv?: never; replaceWithFile: string };
+
+type ReplacementTarget =
+  | { stringToReplace: string; regexToReplace?: never }
```

Comment:
Why carry a string and a regex, when a single property would suffice?

Reply:
exact-match-on-strings is simpler, less error-prone, and require no escaping beyond what json requires for `\`

considering the common scenarios like emails and urls, `user@company.com` or `https://foo.com/api`, it's gonna result in more success for users.

Reply:
@mshanemc Still a bit confused. Can you share some examples in json where expressing a value as a string is simpler than a regex.